packages/monaco/src/browser/monaco-semantic-highlighting-service.ts

```diff
@@ -32,9 +33,51 @@ export class MonacoSemanticHighlightingService extends SemanticHighlightingServi
     @inject(EditorManager)
     protected readonly editorManager: EditorManager;
 
+    @inject(MonacoEditorService)
+    protected readonly monacoEditorService: MonacoEditorService;
+
     protected readonly decorations = new Map<string, Set<string>>();
     protected readonly toDisposeOnEditorClose = new Map<string, Disposable>();
 
+    // laguage id -> (scope index -> model decoration)
+    protected readonly modelDecorations = new Map<string, Map<number, monaco.editor.IModelDecorationOptions>>();
+    protected readonly decorationTypeKeys = new IdGenerator('DecorationType');
+
+    register(languageId: string, scopes: string[][] | undefined): Disposable {
+        const result = super.register(languageId, scopes);
+        if (scopes) {
+            const map = new Map<number, monaco.editor.IModelDecorationOptions>();
+            for (let index = 0; index < scopes.length; index++) {
+                const modelDecoration = this.toModelDecoration(scopes[index]);
+                if (modelDecoration) {
+                    map.set(index, modelDecoration);
+                }
+            }
+            this.modelDecorations.set(languageId, map);
+        }
+        // TODO(nridge):
+        //  1. Remove the decoration types.
+        //  2. Update the decoration types if the theme changes.
+        return result;
+    }
+
+    protected toModelDecoration(scopes: string[]): monaco.editor.IModelDecorationOptions | undefined {
+        // TODO: why for-of? How to pick the right scope? Is it fine to get the first element (with the narrowest scope)?
+        const key = this.decorationTypeKeys.nextId();
+        for (const scope of scopes) {
+            const tokenTheme = this.tokenTheme();
+            const metadata = tokenTheme.match(undefined, scope);
+            const colorIndex = monaco.modes.TokenMetadata.getForeground(metadata);
+            const color = tokenTheme.getColorMap()[colorIndex];
+            const options: monaco.editor.IDecorationRenderOptions = {
+                color: color.toString(),
+            };
+            this.monacoEditorService.registerDecorationType(key, options);
```

Comment:
We should make remove all decoration types if a disposable returned by `register` is disposed.

Reply:
Done